src/prefixtrie.cpp

```diff
@@ -0,0 +1,486 @@
+    auto& children = node->children;
+    auto it = children.lower_bound(key);
+    if (it != children.end()) {
+        if (it->first == key)
+            return it->second;
+        count = match(key, it->first);
+    }
+    if (count == 0 && it != children.begin()) {
+        --it;
+        count = match(key, it->first);
+    }
+    if (count == 0) {
+        ++size;
+        it = children.emplace(key, std::make_shared<Node>()).first;
+        it->second->data = std::make_shared<TData>();
+        return it->second;
+    }
+    if (count < it->first.size()) {
+        const TKey prefix(key.begin(), key.begin() + count);
+        const TKey postfix(it->first.begin() + count, it->first.end());
+        auto nodes = std::move(it->second);
+        children.erase(it);
+        ++size;
+        it = children.emplace(prefix, std::make_shared<Node>()).first;
+        it->second->children.emplace(postfix, std::move(nodes));
+        it->second->data = std::make_shared<TData>();
+        if (key.size() == count)
+            return it->second;
+    }
+    return insert(TKey(key.begin() + count, key.end()), it->second);
+}
+
+template <typename TKey, typename TData>
+void CPrefixTrie<TKey, TData>::erase(const TKey& key, std::shared_ptr<Node>& node)
+{
+    std::vector<typename TChildren::value_type> nodes;
+    nodes.emplace_back(TKey(), node);
+    using CBType = callback<std::shared_ptr<Node>>;
+    CBType cb = [&nodes](const TKey& k, std::shared_ptr<Node> n) {
+        nodes.emplace_back(k, n);
+    };
+    if (!find(key, node, cb))
+        return;
+
+    nodes.back().second->data = std::make_shared<TData>();
+    for (; nodes.size() > 1; nodes.pop_back()) {
+        // if we have only one child and no data ourselves, bring them up to our level
+        auto& cNode = nodes.back().second;
+        auto onlyOneChild = cNode->children.size() == 1;
+        auto noData = cNode->data->empty();
+        if (onlyOneChild && noData) {
+            auto child = cNode->children.begin();
+            auto& prefix = nodes.back().first;
+            auto newKey = prefix;
+            auto& postfix = child->first;
+            newKey.insert(newKey.end(), postfix.begin(), postfix.end());
+            auto& pNode = nodes[nodes.size() - 2].second;
+            pNode->children.emplace(newKey, std::move(child->second));
+            pNode->children.erase(prefix);
+            --size;
+            continue;
+        }
+
+        auto noChildren = cNode->children.empty();
+        if (noChildren && noData) {
+            auto& pNode = nodes[nodes.size() - 2].second;
+            pNode->children.erase(nodes.back().first);
+            --size;
+            continue;
+        }
+        break;
+    }
+}
+
+template <typename TKey, typename TData>
+CPrefixTrie<TKey, TData>::CPrefixTrie() : size(0), root(std::make_shared<Node>())
+{
+    root->data = std::make_shared<TData>();
+}
+
+template <typename TKey, typename TData>
+template <typename TDataUni>
```

Comment:
TDataUni allows us to have both lvalue and rvalue data inputs on this method. The alternative is to have two overloads of this method.